tests/components/mqtt/test_init.py

```diff
@@ -1288,15 +1308,253 @@ def _callback_b(msg: ReceiveMessage) -> None:
     await hass.async_block_till_done()
     await mqtt.async_subscribe(hass, "test/state", _callback_b, qos=1)
     async_fire_mqtt_message(
-        hass, "test/state", "online"
-    )  # Simulate a (retained) message replaying
+        hass, "test/state", "online", qos=0, retain=False
+    )  # Simulate an other non retained message after the second subscription
     async_fire_time_changed(hass, utcnow() + timedelta(seconds=1))
     await hass.async_block_till_done()
     async_fire_time_changed(hass, utcnow() + timedelta(seconds=1))
     await hass.async_block_till_done()
+    # Both subscriptions should receive updates
+    assert len(calls_a) == 1
+    assert len(calls_b) == 1
+    mqtt_client_mock.subscribe.assert_called()
+
+
+@patch("homeassistant.components.mqtt.client.INITIAL_SUBSCRIBE_COOLDOWN", 0.0)
+@patch("homeassistant.components.mqtt.client.DISCOVERY_COOLDOWN", 0.0)
+@patch("homeassistant.components.mqtt.client.SUBSCRIBE_COOLDOWN", 0.0)
+async def test_replaying_payload_same_topic(
+    hass: HomeAssistant,
+    mqtt_client_mock: MqttMockPahoClient,
+    mqtt_mock_entry: MqttMockHAClientGenerator,
+) -> None:
+    """Test replaying retained messages.
+
+    When subscribing to the same topic again, SUBSCRIBE must be sent to the broker again
+    for it to resend any retained messages for new subscriptions.
+    Retained messages must only be replayed for new subscriptions, except
+    when the MQTT client is reconnection.
+    """
+    mqtt_mock = await mqtt_mock_entry()
+
+    # Fake that the client is connected
+    mqtt_mock().connected = True
+
+    calls_a: list[ReceiveMessage] = []
+    calls_b: list[ReceiveMessage] = []
+
+    def _callback_a(msg: ReceiveMessage) -> None:
+        calls_a.append(msg)
+
+    def _callback_b(msg: ReceiveMessage) -> None:
+        calls_b.append(msg)
+
+    await mqtt.async_subscribe(hass, "test/state", _callback_a)
+    async_fire_mqtt_message(
+        hass, "test/state", "online", qos=0, retain=True
+    )  # Simulate a (retained) message played back
+    await hass.async_block_till_done()
+    assert len(calls_a) == 1
+    mqtt_client_mock.subscribe.assert_called()
+    calls_a = []
+    mqtt_client_mock.reset_mock()
+
```

Comment:
Missing test.
async_subscribe for "callback B" has been done ('callback B' subscribe is pending)
SUBSCRIBE_COOLDOWN is in progress, so SUBSCRIBE to MQTT server has NOT been sent.
RETAIN=False message is sent from the server.

SPEC behavior says calls_a = 1, calls_b = 0
CURRENT behavior is calls_a = 1, calls_b = 1

Reply:
As we did not resubscribe to the broker does not mean we are not subscribed. So yes we get a reply with retain = 0, but only if there is anything published in the mean time. That data will be live, and it should not have a retain flag. After the resubscribe was done we are not interested in any stale retained info on that topic if we have a more actual value. Goal is to get the latest value available (live) or a retained value (last known retained). The only thing we miss is that we do not let the retained value to override any received more actual value.

Reply:
In this scenario, callback B **never subscribed**.  There is no 'resubscribe', callback B was not yet subscribed yet.  My point was that callback B generated a SUBSCRIBE call to be sent to the MQTT server.  And that has not yet been sent.  So, logically, callback B has not yet been subscribed, though callback A has.
If callback A and B used different topics, callback B would not receive or process ANY data until the SUBSCRIBE went to the MQTT server.  But because they happen to be the same topic, you are allowing callback B to receive/process messages before the corresponding SUBSCRIBE has been sent to the server.

But ignoring that.  A RETAIN=1 message **can not be stale** (using your assumptions).

Any client that would use RETAIN=1 messages, will use RETAIN=1 on **ALL** messages if they always send full state as you assume.  The ONLY reason a client would combine RETAIN=1 and RETAIN=0 messages, is if the RETAIN=0 messages DID NOT contain the entire state.  Which you seem to think impossible.  Despite me giving you a sample program doing just that, which was a valid use case.  But once again, you have a "Well, it doesn't happen in *my* environment, therefore we should not support it" response (of course, everyone else's environment is just like yours, right?)

So, if we assume the client is either always sending RETAIN=1, then any RETAIN=1 message you receive will be an exact duplicate of the previous RETAIN=0 message received on the same topic.  **not stale**.  Why? Because the client originally sent that RETAIN=0 message as RETAIN=1, because they're sending EVERY message as RETAIN=1.

Now, as I said, it may be the desired behavior that callback B gets messages before it the SUBSCRIBE call is sent to the MQTT server, because callback A happens to already be subscribed on the same topic.  But if that's the case, there is no reason to actually send the SUBSCRIBE for callback B to the server if it receives/processes ANY message between when you added that topic to the pending list, and when the pending list is cleared.  At best, all that SUBSCRIBE call is going to do is trigger a new message (RETAIN=1) you are going to ignore.

Reply:
When a second subscription comes in, asuming the re-subscription to the broker comes later, will (if retained messages were send and stored) lead to the behavior that those messages will be replayed. It is only an edge case that within the debounce period a messages was received before resubscribing. We do not even know if that message was retained. But we do have an actual state for the topic at that time, as soon as the debouncer times out any or none retained message will come. But those will not be newer that the latest message received. May be they are even older, because the last payload reived by the broker was not retained, but the broker also had a stored retained payload. In that case , that payload will be replayed. What I say is, that we should ignore that after we had an actual update (retain flag=0).